### PR TITLE
ENH: Add examples to the template.

### LINKED ISSUE #34: Add `examples` folder content to the module template
See:

  https://github.com/KitwareMedical/ITKMorphologicalContourInterpolation/tree/master/examples

for what the content can be like.

### PR DESCRIPTION
**Pull request by jhlegarreta**
Add a hint to foster writing examples for the contributed classes.

Resolves #34.

### CODE REVIEW

**Comment by jhlegarreta** [comment]
I've further thinking about this. In absence of any elaborated examples, a fairly easy solution, and a nice and much needed addition could be to write a **performance benchmark**. Considering the way the [ITKPerformanceBenchmarking](https://github.com/InsightSoftwareConsortium/ITKPerformanceBenchmarking/) is structured, an example reporting the performance could fit into that framework. Otherwise, if we think that performance benchmarks should be made more explicit and deserve a separate module, we may need to further think how we can propose this. In either case, I'd be happy to work on that.

File: {{cookiecutter.project_name}}/examples/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+cmake_minimum_required(VERSION 3.10.2)
+project({{ cookiecutter.module_name }}Examples)
+
+find_package(ITK REQUIRED
+  COMPONENTS
+    ITKImageIO
```
Comment (dzenanz) [inline {{cookiecutter.project_name}}/examples/CMakeLists.txt]: This is problematic if the example is built as part of ITK. See [this](https://github.com/InsightSoftwareConsortium/ITKMontage/commit/3cca40a598e359cd28bc15ed5903db9a0100886b).

Comment (dzenanz) [reply {{cookiecutter.project_name}}/examples/CMakeLists.txt]: @fbudin69500 do you think we should add this to the module template?

File: {{cookiecutter.project_name}}/examples/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+cmake_minimum_required(VERSION 3.10.2)
+project({{ cookiecutter.module_name }}Examples)
```
Comment (jhlegarreta) [inline {{cookiecutter.project_name}}/examples/CMakeLists.txt]: Do we need to add `CXX` or does this help CMake somehow exclude any Python example file that is added along?

Comment (jhlegarreta) [inline {{cookiecutter.project_name}}/examples/CMakeLists.txt]: Also, do we need to enable and add tests for the examples?

Comment (dzenanz) [reply {{cookiecutter.project_name}}/examples/CMakeLists.txt]: If the example does something, we can use it to write a test, see [this](https://github.com/KitwareMedical/ITKMorphologicalContourInterpolation/blob/dcae5d5b693098a98236d21e5191baa2f8abf74b/examples/CMakeLists.txt#L18-L23). Having test data in there is harder to properly support, that's why [I removed it](https://github.com/KitwareMedical/ITKMorphologicalContourInterpolation/commit/716b4a3ccbb329e54d9e2bd1d1d914699e1ccb1b).

File: {{cookiecutter.project_name}}/examples/CMakeLists.txt (hunk 1 above)
Comment (fbudin69500) [reply {{cookiecutter.project_name}}/examples/CMakeLists.txt]: This seems difficult to address cleanly for now as the template would have to explicitly instantiate certain IOs and it is not possible to know ahead of time which IOs should be instantiated. There could be hints in the CMakeLists.txt with commented out lines of code with the IO names that the developer could either uncomment or remove, but that doesn't seem ideal. @dzenanz : Was an issue created in the issue tracker for that Meta module IO issue?

Comment (dzenanz) [reply {{cookiecutter.project_name}}/examples/CMakeLists.txt]: I just created an issue: https://github.com/InsightSoftwareConsortium/ITK/issues/359

File: {{cookiecutter.project_name}}/examples/CMakeLists.txt (hunk 2 above)
Comment (thewtex) [reply {{cookiecutter.project_name}}/examples/CMakeLists.txt]: Python examples don't need to be built -- the default include `CXX` -> this is good.

**Comment by thewtex** [comment]
Awesome @jhlegarreta ! This is a great contribution. Examples are essential.

For the testing data, perhaps we can use the [CMake FetchContent module](https://cmake.org/cmake/help/latest/module/FetchContent.html#fetch-content-examples)? It requires CMake 3.11, but we could just add `cmake_minimum_required(VERSION 3.11)` in *examples/CMakeLists.txt* only.

File: {{cookiecutter.project_name}}/examples/CMakeLists.txt (hunk 2 above)
Comment (jhlegarreta) [reply {{cookiecutter.project_name}}/examples/CMakeLists.txt]: :thinking:  So do we need to use
```
project({{ cookiecutter.module_name }}Examples CXX)
```
explicitly then?

**Comment by jhlegarreta** [comment]
> Awesome @jhlegarreta ! This is a great contribution. Examples are essential.
> 
> For the testing data, perhaps we can use the [CMake FetchContent module](https://cmake.org/cmake/help/latest/module/FetchContent.html#fetch-content-examples)? It requires CMake 3.11, but we could just add `cmake_minimum_required(VERSION 3.11)` in _examples/CMakeLists.txt_ only.

:thinking: Not sure how the CMake `FetchContent` module plays a role here or how does it fit with having tests for the examples.

**Comment by thewtex** [comment]
> Not sure how the CMake `FetchContent` module plays a role here or how does it fit with having tests for the examples.

FetchContent can be used inside an example so the example can be copied outside the module as a starting point for a new independent application. Configuration with FetchContent is much simpler than CMake ExternalData, so the example can remain simple.

**Comment by hjmjohnson** [comment]
@jhlegarreta This is more than a year out of date.  Can it be abandoned?

**Comment by jhlegarreta** [comment]
You're right Hans, it's been a while. I have not found the time to investigate more.

IMHO, having the `ITKModuleTemplate` a mechanism to make room for examples is useful, since we will want remotes to be as complete as possible. I do not foresee that I will have time to pick this up any time soon, so if you feel like it should be abandoned, go ahead.

If at some point someone has the cycles to tackle this, unless we delete the branch, they will be able to re-open it or to get some inspiration from this.

**Comment by hjmjohnson** [comment]
@dzenanz Can you please re-address your concerns with this issue?  If it is an incremental improvement with known deficencies, could it be merged?  If not, let's make it into an issue and close this PR.

**Comment by dzenanz** [comment]
It is not a simple fix. I updated the issue #34.

File: {{cookiecutter.project_name}}/examples/CMakeLists.txt (hunk 1 above)
Comment (dzenanz) [reply {{cookiecutter.project_name}}/examples/CMakeLists.txt:6]: We could apply solution from `ITKMontage`:
https://github.com/InsightSoftwareConsortium/ITKMontage/blob/v0.8.2/examples/CMakeLists.txt#L4-L22

**Comment by dzenanz** [comment]
This PR was fairly close to working. Applying solution from https://github.com/InsightSoftwareConsortium/ITKMontage/blob/v0.8.2/examples/CMakeLists.txt#L4-L22, along with a general update should make it work. Jon, can you revive it?

**Comment by jhlegarreta** [comment]
> This PR was fairly close to working. Applying solution from https://github.com/InsightSoftwareConsortium/ITKMontage/blob/v0.8.2/examples/CMakeLists.txt#L4-L22, along with a general update should make it work. Jon, can you revive it?

PR #163.

File: {{cookiecutter.project_name}}/examples/CMakeLists.txt (hunk 1 above)
Comment (jhlegarreta) [reply {{cookiecutter.project_name}}/examples/CMakeLists.txt:6]: PR #163.

**Pull request closed without merging**